### PR TITLE
964 search results should respect authorization

### LINKED ISSUE #964: Search results should respect Authorization
Currently (on ontohub.org), i can search for ontologies which are part of of private repositories even though i do not have access to them (i'm not logged in).

Here is an example:
- general [search](https://ontohub.org/ontologies) with **DMB-CommonSource** in the search field.

Results in one result. However that ontology is in a private repository (and if you click on the link you are redirected to a 'you are not authorized'-page).

This should **not** be fixed in the current version of the search, but should instead be considered when *migrating* to elasticsearch (that is why i have assigned @ebolloff for now).

### LINKED ISSUE #1640: Adjust ontology visibility on repository access change
When I have a publicly readable repository with some ontologies
And I set its access to be privately readable
Then the ontologies should not be visible in the site-wide ontology search for the public.

But they are...
Maybe they need to be re-indexed?

### PR DESCRIPTION
**Pull request by ebolloff**
This should fix #964 and fix #1640.

### CODE REVIEW

**Comment by ebolloff** [comment]
I'll review the commits of @eugenk and he'll review mine.

File: app/controllers/home_controller.rb [hunk 1]
```diff
@@ -5,7 +5,8 @@ class HomeController < ApplicationController
 
   def index
     @comments = Comment.latest.limit(10).all
-    @versions = OntologyVersion.latest.where(state: 'done').limit(10).all
+    @versions = OntologyVersion.accessible_by(current_user).latest.
+      where(state: 'done').limit(10).all
```
Comment (eugenk) [inline app/controllers/home_controller.rb]: I think (not sure, though) that the conversion to an array with`.all` is unnecessary overhead.

File: features/OntologySearch.feature [hunk 2]
```diff
@@ -124,4 +124,18 @@ Scenario: I want to search for a not existing ontology in a repository
 Scenario: I want to search for a ontology with a category
   Given there is an ontology with a category
   When I open the ontologies overview page
-  Then I should see all ontologies with a category
\ No newline at end of file
+  Then I should see all ontologies with a category
+
+Scenario: I want to search for a ontology I have access to
+  Given I am logged in as the owner of a private repository
+  And a there is a ontology in this repository
```
Comment (eugenk) [inline features/OntologySearch.feature]: typo: "And a there is *an* ontology"

File: features/OntologySearch.feature [hunk 3]
```diff
@@ -124,4 +124,18 @@ Scenario: I want to search for a not existing ontology in a repository
...
+  When I open the ontologies overview page
+  And I search for this ontology
+  Then I should see the ontology in this repo
```
Comment (eugenk) [inline features/OntologySearch.feature]: Why do you abbreviate repository (only) here?

File: features/OntologySearch.feature [hunk 4]
```diff
@@ -124,4 +124,18 @@ Scenario: I want to search for a not existing ontology in a repository
...
+
+Scenario: I want to search for a ontology I don't have access to
+  Given I am not logged in as the owner of a private repository
+  And a there is a ontology in this repository
+  When I open the ontologies overview page
+  And I search for this ontology
+  Then I shouldnt see the ontology in this repo
```
Comment (eugenk) [inline features/OntologySearch.feature]: typo: "I shouldn**'**t"

File: features/step_definitions/ontology_search_steps.rb [hunk 5]
```diff
@@ -217,3 +217,34 @@
...
+  login_as @user, :scope => :user
+end
+
+Given(/^a there is a ontology in this repository$/) do
+  @ontology = FactoryGirl.create :ontology, repository: @repo
+  @repo.ontologies << @ontology
+end
+
+Then(/^I should see the ontology in this repo$/) do
+  page.should have_content(@ontology.name)
```
Comment (eugenk) [inline features/step_definitions/ontology_search_steps.rb]: Please use expect syntax: `expect(page). to have_content(@ontology.name)`

File: features/step_definitions/ontology_search_steps.rb [hunk 6]
```diff
@@ -217,3 +217,34 @@
...
+end
+
+Then(/^I shouldnt see the ontology in this repo$/) do
+  page.should_not have_content(@ontology.name)
```
Comment (eugenk) [inline features/step_definitions/ontology_search_steps.rb]: Please use expect syntax

File: spec/factories/repository_factory.rb [hunk 7]
```diff
@@ -43,5 +43,14 @@
       remote_type { 'mirror' }
       source_type { 'git' }
     end
+
+    factory :private_repository do |repository|
+      repository.after(:build) do |repository|
+        repository.access = 'private_rw'
```
Comment (eugenk) [inline spec/factories/repository_factory.rb]: Please move the access up a line (this is a model attribute which can be set directly):
```ruby
    factory :private_repository do |repository|
      access { 'private_rw' }
      repository.after(:build) do |repository|
        editor = create :user
        ```

**Comment by eugenk** [comment]
This works like a charm. After you worked off the comments, I will sign this off.

**Comment by ebolloff** [comment]
:+1:

**Comment by eugenk** [comment]
:+1:

**Pull request merged**